packages/block-editor/src/components/block-edit/index.js

```diff
@@ -20,6 +24,76 @@ import { BlockEditContextProvider, useBlockEditContext } from './context';
  */
 export { useBlockEditContext };
 
+function BlockControlFilters( props ) {
```

Comment:
Isn't this the same thing as just adding an empty component and wrapping it with `withFilters` or something like that?

Reply:
I guess there are a few differences:

 - The filter doesn't have to render the previous value. This is a great improvement DevX wise.
 - It adds a check `shouldDisplayControls`: I'm guessing that's to improve performance and avoid rendering these controls if not necessary. While this is a good thing probably, it does add a bit of overhead in the sense that most of these filters actually render `InspectorControls` or `BlockControls` that do these checks as well. So using a context value for these checks might still be a good idea. Also, should we support the case where we pass flag to enable `__experimentalExposeControlsToChildren` behavior or not? (so let's say that this is generally a good thing)
 - The other thing is that folks can render anything in this filter and not just controls, like random Divs... So maybe, we should forbid that by adding some hidden wrapper or something like that or finding a way to check that it renders no DOM nodes.

Reply:
For the performance aspect, we should be a bit careful here to not actually degrade performance. Some existing hooks might have existed earlier (like by checking a block support) while here we first check `shouldDisplayControls` then check the block supports ... within the hook. So I wonder if we should first land `shouldDisplayControls` as a context value before this PR.

Reply:
> While this is a good thing probably, it does add a bit of overhead in the sense that most of these filters actually render InspectorControls or BlockControls that do these checks as well. So using a context value for these checks might still be a good idea.

The overhead is almost negligible. I'd prefer keeping all this internal instead of exposing it through context, although this context in particular can be entirely internal if we don't use the existing context.

> Also, should we support the case where we pass flag to enable __experimentalExposeControlsToChildren behavior or not? (so let's say that this is generally a good thing)

It supports it. The filters will render.

> For the performance aspect, we should be a bit careful here to not actually degrade performance. Some existing hooks might have existed earlier (like by checking a block support) while here we first check shouldDisplayControls then check the block supports ... within the hook. So I wonder if we should first land shouldDisplayControls as a context value before this PR.

Not sure how this would degrade performance. We still guard hasSelectedInnerBlock with a __experimentalExposeControlsToChildren check. Using context for InspectorControls and BlockControls would be better yes.

Reply:
> The filter doesn't have to render the previous value. This is a great improvement DevX wise.

That's correct. If this is a good pattern for other filters, at some point we could move this to the components package next to `withFilters`, but let's see how it goes.

Reply:
> It supports it. The filters will render.

I mean some controls won't render if the block itself is not selected and that additional check and rendering is not necessary in this case.